Added css to css module script

somya51p opened the pull request:
@jennydaman I have created the script to convert CSS to CSS modules using regular expressions in js. Please review the PR and suggest changes if any.

sherrif10 commented:
@somya51p  What exactly does this do. How can i test this generally.

somya51p commented:
> @somya51p What exactly does this do. How can i test this generally.

Hey, @sherrif10  It's associated with issue #167. This script would automatically convert CSS files to CSS modules.

jennydaman commented on scripts/cssToCssModule.js:
```diff
@@ -0,0 +1,79 @@
+const fs = require("fs");
+const path = require("path");
+
+const blackListedClasses = [
```
these are from patternfly right?

somya51p replied:
> these are from patternfly right?

Yes, both bootstrap and patternfly

jennydaman commented on scripts/cssToCssModule.js:
```diff
@@ -0,0 +1,79 @@
+    fs.rename(cssFileName, cssModuleName, () => {});
+    let data = textData.replace(
+      `import './${cssFileName.split("/").slice(-1)}'`,
+      `import styles from './${cssModuleName.split("/").slice(-1)}'`
```
What happens if multiple stylesheets are imported in the same JSX file? (Or how can you verify that this is not a concern?)

somya51p replied:
> What happens if multiple stylesheets are imported in the same JSX file? (Or how can you verify that this is not a concern?)

I need to check that for such cases. As in the current codebase, such a thing does not persist so it would work fine ig. The script would convert the files to CSS modules successfully but I cant say about the impact at this very moment.

jennydaman replied:
Okay.

It would be helpful if you could provide a way to _verify_ this is not a concern, i.e., explain how you did your "analysis."

I looked into this and here's a script I ran:

```shell
fd -e jsx -x bash -c '[ $(grep -F .css {} | grep -v node_modules | wc -l) -lt 2 ] || echo "WARNING: {} imports more than 1 local CSS file."' src
```

This script translates to: "for all JSX files under the `src/` directory, count the number of lines which include the substring ".css" but do not include the substring "node_modules", and if this number is greater than or equal to 2, print a warning."

somya51p replied:
> Okay.
> 
> It would be helpful if you could provide a way to _verify_ this is not a concern, i.e., explain how you did your "analysis."
> 
> I looked into this and here's a script I ran:
> 
> ```shell
> fd -e jsx -x bash -c '[ $(grep -F .css {} | grep -v node_modules | wc -l) -lt 2 ] || echo "WARNING: {} imports more than 1 local CSS file."' src
> ```
> 
> This script translates to: "for all JSX files under the `src/` directory, count the number of lines which include the substring ".css" but do not include the substring "node_modules", and if this number is greater than or equal to 2, print a warning."

Ok, will take care of that, and also write unit tests. Thanks for the review.

jennydaman commented on scripts/cssToCssModule.js:
```diff
@@ -0,0 +1,79 @@
+              if (blackListedClasses.includes(cls)) {
+                  res += cls;
+              } else {
+                  res += `\${styles[${  cls  }]}`;
```
do we need quotes around cls?

somya51p replied:
> do we need quotes around cls?

don't need that for normal `className`

jennydaman replied:
During our meeting we talked about how we do need quotes here. I suggest you write examples and a unit test for this case.

jennydaman reviewed (commented):
Disclaimer: I haven't tried running this yet, I will try later.

somya51p commented:
@jennydaman Please look into the changes made and tests created.

somya51p commented:
Hey @jennydaman, to be precise here are the changes that I had made in the latest commits of this PR. 
1. Created helper functions and added comments.
2. Added quotes for single classNames. 
3. Removed bug for multiple blacklisted classes.too
4. Wrote tests for all the different test cases possible.

Please review it and let me know if I am missing out something. Thankyou!

jennydaman commented on src/scripts/cssToCssModule.js:
```diff
@@ -0,0 +1,162 @@
+        }
+    });
+    return arrayOfFiles;
+};
```
Here's how I would've done it:

```javascript
const fs = require('fs');
const path = require('path');

/**
 * @param {string} p an existing path
 * @returns list of all subpaths under p, including itself
 */
function findAllPaths(startingPath) {
  if (fs.statSync(startingPath).isDirectory()) {
    const subpaths = fs.readdirSync(startingPath).map((subpath) => {
      return findAllPaths(path.join(startingPath, subpath));
    });
    return Array.prototype.concat.apply([startingPath], subpaths);
  }
  return [startingPath];
}

// this helper function here has BOTH a self-explanatory name AND a simple implementation,
// so I am choosing to not add documentation
function filterByFileExtension(paths, suffix) {
  return paths.filter((p) => {
    return fs.statSync(p).isFile() && p.endsWith(suffix);
  });
}

// example

const exampleDirectory = 'ChRIS_store_ui/src';

filterByFileExtension(findAllPaths(exampleDirectory), '.jsx').forEach(p => {
  console.log(p);
});
```

I've identified the individual units of functionality of this task and created a helper for each one. This way, each function is individually testable, so if either one is doing the wrong thing, it's much easier to identify where the bug is in my code, and also the fix for my code will be much smaller. It can also be easier for others to read, understand, and review my code, because the names of my functions are descriptive and there is docstring where necessary.

jennydaman reviewed (changes requested):
Hi @somya51p 

I am still overwhelmed with a backlog of work but I see that you have worked very hard on this so I am providing some initial feedback now on this weekend. I will add more comments later.

somya51p commented:
> Hi @somya51p
> 
> I am still overwhelmed with a backlog of work but I see that you have worked very hard on this so I am providing some initial feedback now on this weekend. I will add more comments later.

Ohk, Thanks a lot for taking out time and reviewing it :)
I will study your comments and implement them in the next commit.
